## Start with cProfilling

In [2]:
import cProfile
import pstats

def sample_function():
    # Some code here to profile
    sum = 0
    for i in range(10000):
        sum += i
    return sum

# Profiling the function
cProfile.run('sample_function()', 'profile_stats')

# Reading the stats
p = pstats.Stats('profile_stats')
p.sort_stats('cumulative').print_stats(10)  # prints the top 10 functions where time was spent


Wed Dec  6 09:06:54 2023    profile_stats

         4 function calls in 0.001 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.001    0.001 {built-in method builtins.exec}
        1    0.000    0.000    0.001    0.001 <string>:1(<module>)
        1    0.001    0.001    0.001    0.001 /tmp/ipykernel_17206/745881689.py:4(sample_function)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




In [3]:
# example.py
def compute():
    data = [i for i in range(1000000)]
    sum = 0
    for i in data:
        sum += i
    return sum

if __name__ == "__main__":
    compute()


In [2]:
!valgrind --tool=cachegrind python compute.py

==32138== Cachegrind, a high-precision tracing profiler
==32138== Copyright (C) 2002-2017, and GNU GPL'd, by Nicholas Nethercote et al.
==32138== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==32138== Command: python compute.py
==32138== 
==32138== brk segment overflow in thread #1: can't grow to 0x4aae000
==32138== (see section Limitations in user manual)
==32138== NOTE: further instances of this message will not be shown
==32138== 
==32138== I refs:        555,382,806


> *I refs:        555,382,806 --> It's a count of how many instructions were fetched from memory.*

In [11]:
# !valgrind --tool=cachegrind python cache_miss_example.py
!valgrind --tool=cachegrind --cache-sim=yes python cache_miss_example.py

==46604== Cachegrind, a high-precision tracing profiler
==46604== Copyright (C) 2002-2017, and GNU GPL'd, by Nicholas Nethercote et al.
==46604== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==46604== Command: python cache_miss_example.py
==46604== 
--46604-- warning: L3 cache found, using its data for the LL simulation.
==46604== brk segment overflow in thread #1: can't grow to 0x4aae000
==46604== (see section Limitations in user manual)
==46604== NOTE: further instances of this message will not be shown
==46604== 
==46604== I refs:        45,279,686
==46604== I1  misses:       309,482
==46604== LLi misses:        12,256
==46604== I1  miss rate:       0.68%
==46604== LLi miss rate:       0.03%
==46604== 
==46604== D refs:        17,704,202  (12,369,727 rd   + 5,334,475 wr)
==46604== D1  misses:       679,043  (   621,510 rd   +    57,533 wr)
==46604== LLd misses:        49,189  (    20,572 rd   +    28,617 wr)
==46604== D1  miss rate:        3.8% (       5.0%    